File: benchmarks/bench_matmul.c
// bench_matmul.c — AI: Matrix Multiply — self-timed
#include <stdio.h>
#include <windows.h>

long long matmul_sum(int n) {
    long long total = 0;
    for (int i = 0; i < n; i++)
        for (int j = 0; j < n; j++) {
            long long acc = 0;
            for (int k = 0; k < n; k++) {
                long long a = (i + k) % 17;
                long long b = (k * j + 1) % 13;
                acc += a * b;
            }
            total += acc;
        }
    return total;
}

int main(void) {
    LARGE_INTEGER freq, t0, t1;
    QueryPerformanceFrequency(&freq);
    QueryPerformanceCounter(&t0);
    long long result = 0;
    for (int rep = 0; rep < 5000; rep++)
        result = matmul_sum(32);
    QueryPerformanceCounter(&t1);
    double ms = (double)(t1.QuadPart - t0.QuadPart) * 1000.0 / freq.QuadPart;
    printf("%lld\n", result);
    fflush(stdout);
    fprintf(stderr, "COMPUTE_MS=%.3f\n", ms);
    return 0;
}

File: benchmarks/bench_matmul.py
import time, sys
def matmul_sum(n):
    total = 0
    for i in range(n):
        for j in range(n):
            acc = 0
            for k in range(n):
                a = (i + k) % 17
                b = (k * j + 1) % 13
                acc += a * b
            total += acc
    return total

t0 = time.perf_counter()
result = 0
for _ in range(200):
    result = matmul_sum(32)
ms = (time.perf_counter() - t0) * 1000
print(result)
print(f'COMPUTE_MS={ms:.3f}', file=sys.stderr)
